File: Python/463_IslandPerimeter.py
class Solution:
    
    def DFS(self, grid: List[List[int]], x, y) -> int:
    
        if grid[x][y] == 0: return 1
        if grid[x][y] == -1: return 0
        
        perimeter = 0    
        grid[x][y] = -1
        
        # check right
        if y < len(grid[x])-1: perimeter += self.DFS(grid, x, y + 1)
        else: perimeter += 1
        # check left
        if y > 0: perimeter += self.DFS(grid, x, y - 1)
        else: perimeter += 1
        # check up
        if x > 0: perimeter += self.DFS(grid, x - 1, y)
        else: perimeter += 1
        # check down
        if x < len(grid)-1: perimeter += self.DFS(grid, x + 1, y)
        else: perimeter += 1
                
        return perimeter
    
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 1:  return self.DFS(grid, i,j)
                

File: C#/463_IslandPerimeter.cs
public class Solution {
    public int IslandPerimeter(int[][] grid) {
        for (int i = 0; i < grid.Length; i++) {
            for (int j = 0; j < grid[i].Length; j++) {
                if (grid[i][j] == 1) return DFS(grid, i, j);
            }
        }
        return 0;
    }
    private int DFS(int[][] grid, int x, int y) {
        if (grid[x][y] == -1) return 0;
        if (grid[x][y] == 0) return 1;
        
        grid[x][y] = -1;
        int perimeter = 0;
        
        // Check top
        perimeter += x > 0 ? DFS(grid, x - 1 , y): 1;
        // Check bottom
        perimeter += x < grid.Length -1 ? DFS(grid, x + 1 , y): 1;
        // Check left
        perimeter += y > 0 ? DFS(grid, x , y - 1): 1;
        // Check right
        perimeter += y < grid[x].Length - 1 ? DFS(grid, x , y + 1) : 1;
        
        return perimeter;
    }
}
